engine.py

```diff
@@ -485,3 +484,146 @@ def _set_project(self):
+        return panel_id    
+    
+    def register_panel(self, title, bundle, widget_class, *args, **kwargs):
+        """
+        Similar to register_command, but instead of registering a menu item in the form of a
+        command, this method registers a UI panel. The arguments passed to this method is the
+        same as for show_panel().
+        
+        Just like with the register_command() method, panel registration should be executed 
+        from within the init phase of the app. Once a panel has been registered, it is possible
+        for the engine to correctly restore panel UIs that persist between sessions. 
+        
+        Not all engines support this feature, but in for example Nuke, a panel can be saved in 
+        a saved layout. Apps wanting to be able to take advantage of the persistance given by
+        these saved layouts will need to call register_panel as part of their init_app phase.
+        
+        In order to show or focus on a panel, use the show_panel() method instead.
+        
+        :param title: The title of the window
+        :param bundle: The app, engine or framework object that is associated with this panel
+        :param widget_class: The class of the UI to be constructed. This must derive from QWidget.
+        
+        Additional parameters specified will be passed through to the widget_class constructor.
+        """
+        # maya doesn't support the concept of saved panels
+        # may be able to use the panelConfiguration command here.
+        pass        
+            
+    def show_panel(self, title, bundle, widget_class, *args, **kwargs):
+        """
+        Shows a panel in a way suitable for this engine. The engine will attempt to
+        integrate it as seamlessly as possible into the host application. If the engine does 
+        not specifically implement panel support, the window will be shown as a modeless
+        dialog instead.
+        
+        :param title: The title of the window
+        :param bundle: The app, engine or framework object that is associated with this window
+        :param widget_class: The class of the UI to be constructed. This must derive from QWidget.
+        
+        Additional parameters specified will be passed through to the widget_class constructor.
+        """
+        from tank.platform.qt import QtCore, QtGui
+        
+        tk_maya = self.import_module("tk_maya")
+        
+        panel_id = self._generate_panel_id(title, bundle)
+        widget_id = "ui_%s" % panel_id
+        self.log_debug("Begin showing panel %s" % panel_id)
+                                    
+        # create a maya window and layout
+        window = pm.window()
+        self.log_debug("Created window: %s" % window)
+        maya_layout = pm.formLayout(parent=window)
+        self.log_debug("Created layout %s" % maya_layout)
+        
+        if pm.control(widget_id, query=1, exists=1):
```

Comment:
Maybe it would be nice if you explained a bit more what is going on? From what I understand, it seems like you are creating a window that is parented to the activeWindow (self._get_dialog_parent()) and then docking it to a pane. Couldn't this have been created directly in the dock? Does this mean the user will briefly see a window popping and then docking? Is this just how it needs to be done with Qt or is this a limitation of the Maya API?